File: vane-core/src/main/java/org/oddlama/vane/core/LootTable.java
package org.oddlama.vane.core;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;
import org.bukkit.NamespacedKey;
import org.bukkit.inventory.ItemStack;
import org.oddlama.vane.core.functional.Consumer2;

public class LootTable {

    private Map<NamespacedKey, List<LootTableEntry>> possibleLoot = new HashMap<>();

    public LootTable() {}

    public LootTable put(final NamespacedKey key, final LootTableEntry entry) {
        possibleLoot.put(key, List.of(entry));
        return this;
    }

    public LootTable put(final NamespacedKey key, final List<LootTableEntry> entries) {
        possibleLoot.put(key, entries);
        return this;
    }

    public LootTable remove(final NamespacedKey key) {
        possibleLoot.remove(key);
        return this;
    }

    public Map<NamespacedKey, List<LootTableEntry>> possibleLoot() {
        return possibleLoot;
    }

    public List<LootTableEntry> flatCopy() {
        List<LootTableEntry> list = new ArrayList<>();
        possibleLoot.values().forEach(list::addAll);
        return list;
    }

    public void generateLoot(final List<ItemStack> output, final Random random) {
        for (final var set : possibleLoot.values()) {
            for (final var loot : set) {
                if (loot.evaluateChance(random)) {
                    loot.addSample(output, random);
                }
            }
        }
    }

    public ItemStack generateOverride(final Random random) {
        double totalChance = 0;
        final double threshold = random.nextDouble();
        final List<ItemStack> resultContainer = new ArrayList<>(1);
        final var lootList = flatCopy();
        Collections.shuffle(lootList, random);
        for (final var loot : lootList) {
            totalChance += loot.chance;
            if (totalChance > threshold) {
                loot.addSample(resultContainer, random);
            }
            if (!resultContainer.isEmpty()) {
                return resultContainer.get(0);
            }
        }
        return null;
    }

    public static class LootTableEntry {

        public double chance;
        public Consumer2<List<ItemStack>, Random> generator;

        public LootTableEntry(int rarityExpectedChests, final ItemStack item) {
            this(1.0 / rarityExpectedChests, item.clone(), 1, 1);
        }

        public LootTableEntry(int rarityExpectedChests, final ItemStack item, int amountMin, int amountMax) {
            this(1.0 / rarityExpectedChests, item.clone(), amountMin, amountMax);
        }

        public LootTableEntry(double chance, final ItemStack item, int amountMin, int amountMax) {
            this(chance, (list, random) -> {
                final var i = item.clone();
                final var amount = random.nextInt(amountMax - amountMin + 1) + amountMin;
                if (amount < 1) {
                    return;
                }

                i.setAmount(amount);
                list.add(i);
            });
        }

        public LootTableEntry(double chance, Consumer2<List<ItemStack>, Random> generator) {
            this.chance = chance;
            this.generator = generator;
        }

        public void addSample(final List<ItemStack> items, final Random random) {
            generator.apply(items, random);
        }

        public boolean evaluateChance(Random random) {
            return random.nextDouble() < chance;
        }
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/core/LootTable.kt
package org.oddlama.vane.core

import org.bukkit.NamespacedKey
import org.bukkit.inventory.ItemStack
import org.oddlama.vane.core.functional.Consumer2
import java.util.*

/**
 * Runtime loot table extension container for additional generated drops.
 */
class LootTable {
    /** Additional loot entries grouped by source key. */
    private val possibleLoot = mutableMapOf<NamespacedKey, MutableList<LootTableEntry>>()

    /** Registers a single loot entry for a key. */
    fun put(key: NamespacedKey, entry: LootTableEntry): LootTable = apply {
        possibleLoot[key] = mutableListOf(entry)
    }

    /** Registers multiple loot entries for a key. */
    fun put(key: NamespacedKey, entries: MutableList<LootTableEntry>?): LootTable = apply {
        possibleLoot[key] = entries ?: mutableListOf()
    }

    /** Removes all entries for a key. */
    fun remove(key: NamespacedKey): LootTable = apply { possibleLoot.remove(key) }

    /** Returns grouped possible loot entries. */
    fun possibleLoot(): Map<NamespacedKey, List<LootTableEntry>> = possibleLoot

    /** Returns all entries flattened into a single mutable list. */
    fun flatCopy(): MutableList<LootTableEntry> =
        possibleLoot.values.flatMapTo(mutableListOf()) { it }

    /** Appends generated loot samples into an output list. */
    fun generateLoot(output: MutableList<ItemStack?>?, random: Random) {
        possibleLoot.values.forEach { set ->
            set.forEach { loot ->
                if (loot.evaluateChance(random)) loot.addSample(output, random)
            }
        }
    }

    /** Generates a single override item by weighted chance accumulation. */
    fun generateOverride(random: Random): ItemStack? {
        var totalChance = 0.0
        val threshold = random.nextDouble()
        val resultContainer = mutableListOf<ItemStack?>()
        flatCopy().shuffled(random).forEach { loot ->
            totalChance += loot.chance
            if (totalChance > threshold) loot.addSample(resultContainer, random)
            if (resultContainer.isNotEmpty()) return resultContainer[0]
        }
        return null
    }

    /**
     * Single loot generation entry.
     *
     * @param chance selection chance.
     * @param generator sample generator callback.
     */
    class LootTableEntry(val chance: Double, val generator: Consumer2<MutableList<ItemStack?>?, Random?>) {
        /** Creates an entry from expected chest rarity with fixed amount 1. */
        constructor(rarityExpectedChests: Int, item: ItemStack) :
            this(1.0 / rarityExpectedChests, item.clone(), 1, 1)

        /** Creates an entry from expected chest rarity and amount range. */
        constructor(rarityExpectedChests: Int, item: ItemStack, amountMin: Int, amountMax: Int) :
            this(1.0 / rarityExpectedChests, item.clone(), amountMin, amountMax)

        /** Creates an entry with explicit chance and amount range. */
        constructor(chance: Double, item: ItemStack, amountMin: Int, amountMax: Int) : this(
            chance,
            Consumer2 { list, random ->
                val amount = random!!.nextInt(amountMax - amountMin + 1) + amountMin
                if (amount >= 1) list!!.add(item.clone().also { it.amount = amount })
            }
        )

        /** Adds a sampled item to the output list. */
        fun addSample(items: MutableList<ItemStack?>?, random: Random?) = generator.apply(items, random)
        /** Evaluates this entry's random chance. */
        fun evaluateChance(random: Random): Boolean = random.nextDouble() < chance
    }
}
